File: Lista Circular/Circular.js
class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class CircularLinkedList {
    constructor() {
        this.last = null;
    }

    
    _getValue(promptMsg) {
        let value = prompt(promptMsg);
        
        while (value === null || isNaN(parseInt(value))) {
            value = prompt("Entrada inválida. " + promptMsg);
        }
        return parseInt(value);
    }

    

    insertBegin() {
        const value = this._getValue("Ingrese un valor para el inicio:");
        const newNode = new Node(value);
        
        if (this.last === null) {
            this.last = newNode;
            this.last.next = this.last;
        } else {
            newNode.next = this.last.next;
            this.last.next = newNode;
        }
    }
    
    insertEnd() {
        const value = this._getValue("Ingrese un valor para el final:");
        const newNode = new Node(value);

        if (this.last === null) {
            this.last = newNode;
            this.last.next = this.last;
        } else {
            newNode.next = this.last.next;
            this.last.next = newNode;
            this.last = newNode; 
        }
    }
    
    insertRandom() {
        const value = this._getValue("Ingrese un valor:");
        const pos = this._getValue("Ingrese la posicion donde desea insertar:");

        const newNode = new Node(value);
        
        if (this.last === null) {
            if (pos <= 0) {
                this.last = newNode;
                this.last.next = this.last;
                console.log("Lista vacía, insertado al inicio/posición 0.");
            }
            return;
        }
        if (pos <= 0) {
            this.insertBegin(); 
            return;
        }

        let current = this.last.next;
        
        for (let i = 0; i < pos - 1; i++) {
            if (current === this.last) {
                
                break; 
            }
            current = current.next;
        }
        
        newNode.next = current.next;
        current.next = newNode;
        
        if (current === this.last) {
            this.last = newNode; 
        }
    }

   

    deleteBegin() {
        if (this.last === null) {
            console.log("La lista esta vacia");
            return;
        }
        const temp = this.last.next; 
        if (this.last.next === this.last) { 
            this.last = null;
        } else {
            this.last.next = temp.next; 
        }
        console.log(`Elemento ${temp.data} eliminado del inicio.`);
        
    }

    deleteEnd() {
        if (this.last === null) {
            console.log("La lista esta vacia");
            return;
        }
        if (this.last.next === this.last) { 
            this.last = null;
            console.log("Elemento único eliminado.");
            return;
        }
        
        let temp = this.last.next; 
        while (temp.next !== this.last) {
            temp = temp.next;
        }
        

        temp.next = this.last.next; 
        this.last = temp; 
        console.log("Elemento eliminado desde el final");
    }
    
    deleteRandom() {
        if (this.last === null) {
            console.log("La lista esta vacia");
            return;
        }
        const pos = this._getValue("Ingresa la posicion a eliminar:");

        let current = this.last.next;
        let previous = this.last; 

        if (pos === 0) {
            this.deleteBegin();
            return;
        }

        for (let i = 0; i < pos; i++) {
             if (current === this.last.next && i > 0) {
                console.log("La posicion no existe");
                return;
            }
            previous = current;
            current = current.next; 
        }
        
        previous.next = current.next; 
        
        if (current === this.last) {
            this.last = previous; 
        }
        console.log(`Elemento eliminado desde la posicion ${pos}`);
    }


    
    search() {
        if (this.last === null) {
            console.log("La lista esta vacia");
            return;
        }
        const value = this._getValue("Ingrese el valor a buscar:");
        let temp = this.last.next;
        let pos = 0;
        
        do {
            if (temp.data === value) {
                console.log(`Valor ${value} encontrado en la posicion ${pos}`);
                return;
            }
            temp = temp.next;
            pos++;
        } while (temp !== this.last.next);
        
        console.log(`Valor ${value} no encontrado en la lista`);
    }

    display() {
        if (this.last === null) {
            console.log("La lista esta vacia");
            return;
        }
        let temp = this.last.next;
        let output = "Lista Circular: ";
        do {
            output += `${temp.data}-> `;
            temp = temp.next;
        } while (temp !== this.last.next);
        console.log(output + "inicio");
    }
}

function main() {
    const cllJS = new CircularLinkedList();
    let choice = 0;
    while (choice !== 9) {
        console.log("\n*********Main Menu*********");
        console.log("1. Agregar al inicio");
        console.log("2. Agregar al final");
        console.log("3. Agregar en una posicion aleatoria");
        console.log("4. Eliminar desde el inicio");
        console.log("5. Eliminar desde el final");
        console.log("6. Eliminar desde una posicion aleatoria");
        console.log("7. Mostrar");
        console.log("8. Buscar");
        console.log("9. Salir");
        choice = parseInt(prompt("Ingresa tu opcion: "));
        
        switch (choice) {
            case 1:
                cllJS.insertbeg();
                break;
            case 2:
                cllJS.insertend();
                break;
            case 3:
                cllJS.randominsert();
                break;
            case 4:
                cllJS.deletebeg();
                break;
            case 5:
                cllJS.deleteend();
                break;
            case 6:
                cllJS.randomdelete();
                break;
            case 7:
                cllJS.display();
                break;
            case 8:
                cllJS.search();
                break;
            case 9:
                return;
            default:
                console.log("Opcion invalida! Por favor intenta de nuevo.");
        }
    }
}


main();

File: Lista Circular/Circular.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class CircularLinkedList:
    def __init__(self):
        self.last = None

    def insertbegin(self):
        value = int(input("Ingrese un valor: "))
        new_node = Node(value)
        
        if self.last is None:
            self.last = new_node
            self.last.next = self.last
        else:
            new_node.next = self.last.next
            self.last.next = new_node

    def insertend(self):
        value = int(input("Ingrese un valor: "))
        new_node = Node(value)
        if self.last is None:
            self.last = new_node
            self.last.next = self.last
        else:
            new_node.next = self.last.next
            self.last.next = new_node
            self.last = new_node

    def insertrandom(self):
        value = int(input("Ingrese un valor: "))
        pos = int(input("Ingrese la posicion donde desea insertar: "))
        
        new_node = Node(value)
        if self.last is None:
            if pos <= 0:
                self.last = new_node
                self.last.next = self.last
            else:
                print("Posición inválida en lista vacía.")
                return

        current = self.last.next
        if pos <= 0:
            self.insert_begin()
            return
            
        for i in range(pos - 1):
            if current == self.last:
                break
            current = current.next
        
        new_node.next = current.next
        current.next = new_node
        
        if current == self.last:
            self.last = new_node

    def deletebegin(self):
        if self.last is None:
            print("La lista esta vacia")
            return
        temp = self.last.next
        if self.last.next == self.last:
            self.last = None
        else:
            self.last.next = temp.next
        

    def deleteend(self):
        if self.last is None:
            print("La lista esta vacia")
            return
        if self.last.next == self.last:
            self.last = None
            return
        
        temp = self.last.next
        while temp.next != self.last:
            temp = temp.next
        temp.next = self.last.next
        self.last = temp
        print("Elemento eliminado desde el final")

    def deleterandom(self):
        if self.last is None:
            print("La lista esta vacia")
            return
        current = self.last.next
        previous = self.last
        pos = int(input("Ingresa la posicion: "))
        for i in range(pos):
            if current == self.last.next and i > 0: 
                 print("La posicion no existe")
                 return
            previous = current
            current = current.next
        previous.next = current.next
        if current == self.last:
            self.last = previous
        print(f"Elemento eliminado desde la posicion {pos}")

    def search(self):
        if self.last is None:
            print("La lista esta vacia")
            return
        value = int(input("Ingrese el valor a buscar: "))
        temp = self.last.next
        pos = 0
        while True:
            if temp.data == value:
                print(f"Valor {value} encontrado en la posicion {pos}")
                return
            temp = temp.next
            pos += 1
            if temp == self.last.next:
                break
        print(f"Valor {value} no encontrado en la lista")

    def display(self):
        if self.last is None:
            print("La lista esta vacia")
            return
        temp = self.last.next
        print("Lista Circular: ", end="")
        while True:
            print(f"{temp.data}-> ", end="")
            temp = temp.next
            if temp == self.last.next:
                break
        print("inicio")

def main():
    cll = CircularLinkedList()
    choice = 0
    while choice != 9:
        print("1. Agregar al inicio")
        print("2. Agregar al final")
        print("3. Agregar en posicion especifica")
        print("4. Eliminar desde el inicio")
        print("5. Eliminar desde el final")
        print("6. Eliminar desde posicion especifica")
        print("7. Mostrar lista")
        print("8. Buscar")
        print("9. Salir")

        try:
            opcion = input("Seleccione una opcion: ").strip()
            choice = int(opcion)
            match choice:
                case 1:
                    cll.insertbegin()

                case 2:
                    cll.insertend()
                case 3:
                    cll.insertrandom()
                case 4:
                    cll.deletebegin()
                case 5:
                    cll.deleteend()
                case 6:
                    cll.deleterandom()
                case 7:
                    cll.display()
                case 8:
                    cll.search()


        except ValueError:
            print("❌ Error: Por favor ingrese un número válido")
if __name__ == "__main__":
    main()
